pkg/agent/secondarynetwork/podwatch/controller.go

```diff
@@ -0,0 +1,389 @@
+		kubeClient:         kubeClient,
+		netAttachDefClient: netAttachDefClient,
+		queue:              workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "podcontroller"),
+		podInformer:        podInformer,
+		podLister:          corelisters.NewPodLister(podInformer.GetIndexer()),
+		nodeName:           nodeName,
+		podCache:           podCache,
+		cniServer:          cniServer,
+	}
+
+	podInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    pc.enqueuePod,
+			DeleteFunc: pc.enqueuePod,
+			UpdateFunc: pc.enqueuePodUpdate,
+		},
+		resyncPeriod,
+	)
+	return &pc
+}
+
+func podKeyGet(pod *corev1.Pod) string {
+	return pod.Namespace + "/" + pod.Name
+}
+
+func (pc *PodController) enqueuePod(obj interface{}) {
+	var err error
+	pod, isPod := obj.(*corev1.Pod)
+	if !isPod {
+		podDeletedState, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			klog.ErrorS(err, "Unexpected object received:", obj)
+			return
+		}
+		pod, ok := podDeletedState.Obj.(*corev1.Pod)
+		if !ok {
+			klog.ErrorS(err, "DeletedFinalStateUnknown object is not of type Pod: ", podDeletedState.Obj, pod)
+			return
+		}
+	}
+	podKey := podKeyGet(pod)
+	pc.queue.Add(podKey)
+}
+
+func (pc *PodController) enqueuePodUpdate(prevObj, newObj interface{}) {
+	return
+}
+
+// handleAddUpdatePod handles Pod Add, Update events and updates annotation if required.
+func (pc *PodController) handleAddUpdatePod(obj interface{}) error {
+	var ret bool
+	pod := obj.(*corev1.Pod)
+
+	secondaryNetwork, ret := checkForPodSecondaryNetworkAttachement(pod)
+	if !ret {
+		klog.InfoS("Pod does not have a NetworkAttachmentDefinition", "Pod", klog.KObj(pod))
+		return nil
+	}
+	// get Pod/Container information (container id, namespace, etc.,) from CNI info cache (added by CNI Server CmdAdd upon
+	// successful Pod and primary network creation).
+	// If container information is not available yet in the CNICache, return error to requeue work
+	podCNIInfo := pc.podCache.GetCNIConfigInfoPerPod(pod.Name, pod.Namespace)
+	if len(podCNIInfo) == 0 {
+		return errors.New("CNI Cache Information not found")
+	} else {
+		networklist, err := parsePodSecondaryNetworkAnnotation(secondaryNetwork)
+		if err != nil {
+			return errors.New("secondary network annotation parsing failed")
+		}
+		pc.configureSecondaryNetwork(pod, networklist, podCNIInfo)
+	}
+	return nil
+}
+
+func (pc *PodController) handleRemovePod(key string) error {
+	var cmdArgs *invoke.Args
+
+	pod := strings.Split(key, "/")
+	// Read the CNI info (stored during Pod creation by cniserver) from cache.
+	// Delete CNI info shared in cache for a specific Pod which is getting removed/deleted.
+	podCNIInfo := pc.podCache.GetCNIConfigInfoPerPod(pod[1], pod[0])
+	if len(podCNIInfo) > 0 {
+		for _, containerInfo := range podCNIInfo {
+			// Clean-up IPAM at whereabouts db (etcd or kubernetes API server) for the pod which is getting removed.
```

Comment:
pod -> Pod

Reply:
Could we add a comment saying other interface configurations will be cleaned up automatically when the container network namespace is destroyed?

Reply:
yes. taken care.